oem-ibm: Fixed the COD crash issue

RameshaR45 opened the pull request:
handled the empty license, invalid license and non object json. and also handled try catch for converting to bin to json

RameshaR45 committed:
oem-ibm: Fixed the COD crash issue

handled the empty license, invalid license and non object json.
and also handled try catch for converting to bin to json

Signed-off-by: RameshaR45 <ramesharohit45@gmail.com>

RameshaR45 commented:
i have tested the COD with invalid license and empty license.
and also tested COD with correct Activation code and invalid Activation code.

Pavithrab7 commented on oem/ibm/libpldmresponder/utils.cpp:
```diff
@@ -195,18 +195,47 @@ int writeToUnixSocket(const int sock, const char* buf, const uint64_t blockSize)
+    {
+        error(
+            "convertBinFileToJson: JSON::from_bson failed for {PATH}, error={ERR}",
+            "PATH", path, "ERR", e.what());
```
follow the latest lg2 pattern - https://github.com/openbmc/docs/blob/master/anti-patterns.md#ill-defined-data-structuring-in-lg2-message-strings

RameshaR45 replied:
i have modified as per pattern

Pavithrab7 commented on oem/ibm/libpldmresponder/utils.cpp:
```diff
@@ -230,15 +259,43 @@ void clearLicenseStatus()
+    }
+    catch (const std::exception& e)
+    {
+        error(
```
same here - https://github.com/openbmc/docs/blob/master/anti-patterns.md#ill-defined-data-structuring-in-lg2-message-strings

RameshaR45 replied:
i have modified as per pattern

Pavithrab7 commented on oem/ibm/libpldmresponder/utils.cpp:
```diff
@@ -195,18 +195,47 @@ int writeToUnixSocket(const int sock, const char* buf, const uint64_t blockSize)
+        error(
+            "convertBinFileToJson: JSON::from_bson failed for {PATH}, error={ERR}",
+            "PATH", path, "ERR", e.what());
+        return Json{};
```
return the result?

RameshaR45 replied:
here i am returning a Json{}(empty json) if fails, insted of returning invalid result or not a object.

Pavithrab7 commented on oem/ibm/libpldmresponder/utils.cpp:
```diff
@@ -230,15 +259,43 @@ void clearLicenseStatus()
+    // Guard: check that we got an object and not null/array/primitive
+    if (!data.is_object())
+    {
+        return;
```
do you want to add any log here?

RameshaR45 replied:
i have added necessary logs

ArchanaKakani commented on oem/ibm/libpldmresponder/utils.cpp:
```diff
@@ -225,20 +248,49 @@ void clearLicenseStatus()
+        if (!entry.contains("Id") || !entry["Id"].is_string())
+        {
+            continue;
+        }
```
Can we combine the above two ifs?

RameshaR45 replied:
i have combined two if conditions

ArchanaKakani commented on oem/ibm/libpldmresponder/utils.cpp:
```diff
@@ -225,20 +248,49 @@ void clearLicenseStatus()
+    std::vector<Json> entries;
+    try
+    {
+        entries = data.value("Licenses", emptyList);
```
is there a possibility of failing here as we have already checked if this is a valid object at line 258?

RameshaR45 replied:
even if i check the json is not a object there is still can fail due to key licencse is missing and default type does not match.
hence i kept in try catch not to fail again

ArchanaKakani replied:
I am just trying to improve the code readability. The code logic is not understood easily with all these error handling. These error checks are needed but just see if they can be combined together.

RameshaR45 replied:
i have removed unnecessary error handling

ArchanaKakani commented on oem/ibm/libpldmresponder/utils.cpp:
```diff
@@ -197,16 +197,39 @@ Json convertBinFileToJson(const fs::path& path)
     std::streampos fileSize;
-
-    // Get the file size
+    if (!file.is_open())
```
Can we combine both conditions?
if (!file.is_open()) and if (fileSize <= 0)

RameshaR45 replied:
i can't combine these two conditions as the file checks might fail while file opening

ArchanaKakani replied:
You can open the file with flag ios::ate where it will seek to the end of stream immediately after open. Then file.tellg() will give you file size
And both the conditions can be combined as

if ((!file.is_open()) || ((fileSize=file.tellg()) <= 0))

Do you see any issues with this?

RameshaR45 replied:
i have combined and handled as per request

pointbazaar committed:
fix ci: move to stdexec::inline_scheduler

Move from `exec::inline_scheduler` to `stdexec::inline_scheduler`.

Looks like the definition has been deprecated and is pointing to
`stdexec::inline_scheduler` anyways.

the relevant header
```

namespace exec {
  // A simple scheduler that executes its continuation inline, on the
  // thread of the caller of start().
  using inline_scheduler
    [[deprecated("Use stdexec::inline_scheduler instead")]] = stdexec::inline_scheduler;
} // namespace exec
```

Tested: Inspection only.

Fixes: 10d0b4b7d1498cfd5c3d37edea271a54d1984e41
Change-Id: I1d23cb624e8902c541eeb31bd31155a3207a2da9
Signed-off-by: Alexander Hansen <alexander.hansen@9elements.com>

pointbazaar committed:
fix ci: move to stdexec::inline_scheduler

Move from `exec::inline_scheduler` to `stdexec::inline_scheduler`.

Looks like the definition has been deprecated and is pointing to
`stdexec::inline_scheduler` anyways.

the relevant header
```

namespace exec {
  // A simple scheduler that executes its continuation inline, on the
  // thread of the caller of start().
  using inline_scheduler
    [[deprecated("Use stdexec::inline_scheduler instead")]] = stdexec::inline_scheduler;
} // namespace exec
```

Tested: Inspection only.

Fixes: 10d0b4b7d1498cfd5c3d37edea271a54d1984e41
Change-Id: I1d23cb624e8902c541eeb31bd31155a3207a2da9
Signed-off-by: Alexander Hansen <alexander.hansen@9elements.com>

manojkiraneda commented:
add to approvelist

ArchanaKakani commented on oem/ibm/libpldmresponder/utils.cpp:
```diff
@@ -195,18 +195,32 @@ int writeToUnixSocket(const int sock, const char* buf, const uint64_t blockSize)
+            "convertBinFileToJson: file open failed {PATH} or size {SIZE} invalid",
+            "PATH", path, "SIZE", fileSize);
+        return Json{};
+    }
```
We need to reset the file position to beginning
file.seekg(0, std::ios::beg);

RameshaR45 replied:
yes, i have set to beginning

geissonator commented:
add to approvelist

RameshaR45 committed:
Revert "fix ci: move to stdexec::inline_scheduler"

This reverts commit d108a28aae7d147fb4117c6fcc0450f842a2c5bf.
